specify/specify7
# Allow for integration with Localization frameworks

[Link to Weblate](https://hosted.weblate.org/projects/specify-7/)

- [x] Migrate to typesafe-i18n library
- [x] Pick a localization UI
- [x] Integrate with it
- [ ] Currently, Weblate is listening for strings on the `l10n-improvements` branch. Need to make it look at `production` instead. (in `test.yml` and `push.yml`)
- [x] Configure a webhook to automatically update weblate on localization changes
- [x] Configure automated translation. Try it out on Spanish and French languages.
- [x] Extend this documentation and adapt it for non-developers: https://github.com/specify/specify7/blob/l10n-improvements/specifyweb/frontend/js_src/lib/localization/README.md#syntax-for-string-parameters-and-plurals

To test:
- Open a data set that was created on a different branch that has validation errors and make sure validation messages are displayed properly
- Trigger any number of the following validation errors and make sure the error message is formatted correctly:
  - on invalid boolean: `value "{value:string}" not resolvable to True or False`
  - on invalid decimal: `value "{value:string}" is not a valid decimal value`
  - on invalid value of read-only pick list:
    ```
    "{value:string}" is not a legal value in this picklist field.

    Click on the arrow to choose among available options.
    ```
  - on invalid agent type: `bad agent type: "{badType:string}". Expected one of {validTypes:string}`
  - on skipped required tree rank: `Missing or unmapped required tree parent rank value for "{names:string}".`
  - on pick list value too long: `value from picklist {pickList:string} longer than the max of {maxLength:number|formatted} for field`
  - on field value too long: `value must not have length greater than {maxLength:number|formatted}`
  - on date with invalid year: `date value must contain four digit year: {value:string}`
  - on date with incorrect format: `bad date value: {value:string}. expected: {format:string}`
  - on coordinate in a wrong format: `bad latitude or longitude value: {value:string}`
  - on latitude out of range: `latitude must be between -90 and 90. actual: {value:string}`
  - on longitude out of range: `longitude must be between -180 and 180. actual: {value:string}`
  - on tree node with empty Name, but non empty other column (i.e, Taxon Author) - `this field must be empty if {column:string} is empty`
  - on missing required field: `field is required by upload plan mapping`
  - Create two nodes in the taxon tree with the name `Uploaded`. Both should be at the root level (have to do this though database). WB data set validation should print: `There are multiple "Uploaded" placeholder values in the tree!`
- Go though all front-end pages and make sure localization strings are rendered like expected. Only need to do this for one language. This is just to check for typos. Things to keep a lookout for includes cases when dynamic localization string is printed with something like `{value|string` (i.e, the value wasn't substituted)
- Modify any localization string in Specify 7 on the `l10n-improvements` branch. Should be able to see it in `Weblate` after less than 10 minutes (link: https://hosted.weblate.org/projects/specify-7/)
- Modify an localization string in Weblate (link: https://hosted.weblate.org/projects/specify-7/). In 3 hours after modification, the string should be modified in Specify 7's source code on the `l10n-improvements` branch. If it requires approval to modify a string, message me and I will give you edit permissions



#### Notes for the Weblate demo:
##### Pros:
- If developers change the English translation in the code, Weblate will mark all other translations of that string as in need of review
- Weblate checks for inconsistent punctuation and a [hundred other things](https://docs.weblate.org/en/latest/user/checks.html#check-same) - really cool
- We are hosted for free, but the platform is open source, so if we are not satisfied with the performance we can always move to self-hosting
- Provides stats and audit log. Has a simple permission system. Translators can subscribe for notifications and receive emails when languages that they are familiar with are in need of translation again.
- If I setup it correctly, anyone from the outside could give us suggestions/comments/propose edits. People whom we explicitly approved could get their changes merged into production automatically!
 -Supports automated translation using Google Translate and 20 other localization services. Though, we need to provide our own API key. Fortunately, we quality under Google Translate's free tier.
- A really fancy feature: you can upload screenshots of the application, it will do OCR, identify all the localization strings used on that component and link them to the screenshot so that translators can see context when localizing! Additionally, I setup it up to tell you which places in the code the string is used in and how many times it is used.
- Provides interesting and in-depth stats. See screenshots

##### Cons:
- The documentation is not great. I had to read the source code to understand how it works
- It supports only some file formats and does not provide easy integration with other formats. I had to do some trickery to make it work (using a fake [weblate-localization](https://github.com/specify/specify7/tree/weblate-localization) branch and two GitHub actions that do a two-way sync and file format translation)
- Documentation mentions support for translations coming from all branches, but they don't describe it in a lot of detail so I am still not sure how that would work. For now, Weblate would work with the `production` branch only. That should be sufficient for the time being as we are moving to a more agile dev strategy.

![image](https://user-images.githubusercontent.com/40512816/209415894-db062585-a58a-4639-8d42-e54e1947ce12.png)

![image](https://user-images.githubusercontent.com/40512816/209415967-e2598349-1d09-4df7-8816-8777051855e4.png)

![image](https://user-images.githubusercontent.com/40512816/209416009-330fd7fd-bb7d-41a5-99f2-f09a4dae809f.png)
